clash-ffi/src/Clash/FFI/VPI/Object/Value/Parse.hs

```diff
@@ -0,0 +1,237 @@
+{-|
+Copyright:    (C) 2022 Google Inc.
+License:      BSD2 (see the file LICENSE)
+Maintainer:   QBayLogic B.V. <devops@qbaylogic.com>
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.FFI.VPI.Object.Value.Parse
+  ( parseBinStr
+  , parseOctStr
+  , parseDecStr
+  , parseHexStr
+  , InvalidBitString(..)
+  , ImpreciseBitString(..)
+  ) where
+
+import           Control.Exception (Exception)
+import qualified Control.Monad as Monad (foldM)
+import qualified Control.Monad.IO.Class as IO (liftIO)
+import           Data.Char (toLower)
+import           Data.Typeable (Typeable)
+import           Foreign.C.String (CString)
+import qualified Foreign.C.String as FFI (peekCString)
+import           Foreign.C.Types (CInt)
+import           GHC.Stack (CallStack, HasCallStack, callStack, prettyCallStack)
+import           GHC.TypeNats (KnownNat)
+import           Text.Read (readMaybe)
+
+import           Clash.Class.BitPack
+import           Clash.Sized.BitVector (BitVector)
+import           Clash.XException (deepErrorX)
+
+import           Clash.FFI.Monad (SimCont)
+import qualified Clash.FFI.Monad as Sim
+import           Clash.FFI.VPI.Object.Value.Format (ValueFormat(..))
+
+-- | An exception thrown when the bit string to parse contains characters
+-- which should not appear for the given format.
+--
+data InvalidBitString
+  = InvalidBitString ValueFormat String CallStack
+  deriving anyclass (Exception)
+
+instance Show InvalidBitString where
+  show (InvalidBitString f s c) =
+    mconcat
+      [ "Invalid bit-string for format "
+      , show f
+      , ": "
+      , s
+      , "\n"
+      , prettyCallStack c
+      ]
+
+parseBinStr
+  :: forall n o
+   . HasCallStack
+  => KnownNat n
+  => CInt
+  -> CString
+  -> SimCont o (BitVector n)
+parseBinStr size bin = do
+  str <- IO.liftIO (FFI.peekCString bin)
+  let is = [size - 1, size - 2 .. 0]
+
+  let go acc (i, x) =
+        case x of
+          '0' -> pure (replaceBit i 0 acc)
+          '1' -> pure (replaceBit i 1 acc)
+          'x' -> pure acc
+          'z' -> pure acc
+          _   -> Sim.throw (InvalidBitString BinStrFmt str callStack)
```

Comment:
`replaceBit` is a fairly expensive operation, I think you're better off shifting + reversing in the end. You could then use lookup tables ([like this](https://hackage.haskell.org/package/haskus-binary-1.0/docs/Haskus-Format-Binary-Bits-Reverse.html)) to be fairly speedy. A `TODO` suffices though, no need for premature optimization.

Reply:
This comment would apply to quite a number of parsing functions, so maybe a source note?